src/messages/mod.rs

```diff
@@ -63,7 +82,9 @@ impl MessageWithBytes {
 #[derive(Eq, PartialEq, Clone)]
 pub struct HopMessageWithBytes {
     /// Wrapped signed message.
-    content: SignedRoutingMessage,
+    full_content: Option<SignedRoutingMessage>,
+    /// Peeked SignedRoutingMessage infos
+    peek_msg: PeekedSignedRoutingMessageInfo,
```

Comment:
It seems that `peek_msg` is only relevant when `full_content` is `None`. So to avoid redundant states, I would change this to a single enum:

```rust
enum HopMessageMode { // you can come up with better name
    Peeked(PeekedSignedRoutingMessageInfo),
    Full(SignedRoutingMessage),
}

struct HopMessageWithBytes {
    content: HopMessageMode,
    // ...
}
```

Reply:
That is not really the case, essentially full_content is only there to be taken. it is a cached value.
So when we took that value, the PeekedSignedRoutingMessageInfo should still be there.

Reply:
Then when we take the message out, we can set `content` back to the peeked thing.

Reply:
Does that really make things clearer, we need to add more code? I guess I could rename full_content to `signed_message_cache`?

but it is possible to implement `take` on `HopMessageMode` if you prefer it this way.

Reply:
Well the only thing that's somehow bothering me is we are duplicating information: when `full_content` is `Some`, then the dst location is stored twice. Not saying this is a major issue, but it feels bit dirty.

Reply:
Seem to be an awful lot of complicated code, for something that is I just cached a big value, and otherwise I need the dst.

```rust
#[derive(Eq, PartialEq, Clone)]
enum HopMessageDeserializedContent {
    Peeked(PartialSignedRoutingMessage),
    Full(SignedRoutingMessage),
}

impl HopMessageDeserializedContent {
    fn take_full(&mut self) -> Option<SignedRoutingMessage> {
        let dst = *self.message_dst();
        let value = mem::replace(self, Self::Peeked(PartialSignedRoutingMessage{dst}));
        match value {
            Self::Peeked(_) => None,
            Self::Full(msg) => Some(msg),
        }
    }

    fn message_dst(&self) -> &Location {
        match self {
            Self::Peeked(msg) => &msg.dst,
            Self::Full(msg) => &msg.routing_message().dst,
        }
    }
}
```

Reply:
so i think i'll stick with the less verbose solution for now.